CONTRIBUTING.md

```diff
@@ -29,30 +31,138 @@ and maintained by Google](https://opensource.google.com/docs/cla/#why), not to
+
+## Test development style guide
+
+### Directory Organization
+
+The directory tree is organized as follows:
+
+* `cloudbuild/` contains google cloud build scripts for running virtual
+    routers in containers on [KNE](https://github.com/google/kne)
+* `feature/` contains definition and tests of feature profiles.
+* `internal/` contains packages used by feature profile tests.
+* `proto/`  contains protobuf files for feature profiles.
+* `tools/` contains code used for CI checks.
+* `topologies/` contains the testbed topology definitions.
+* `yang/` contains autogenerated go structures.
+
+### Test Suite Organization
+
+Under `feature/<featurename>/[<sub-feature>/]tests`, each feature group has its own
+subdirectory. For example:
+
+* `feature/bgp/` container for BGP feature profiles and test suites.
+* `feature/bgp/addpath` contains feature profiles  related to BGP.
+* `feature/bgp/addpath/tests` contains test code related to BGP Add Path
+    features.
+* `feature/lldp/` contains feature profiles for ISIS.
+* `feature/lldp/tests` contains test suite related to LLDP.
+* `internal/deviations` contains code which overrides standard tests where
+    there are known bugs.
+
+Under each feature group, the `*.go` file should be named after an appropriate
+[github featureprofiles project](https://github.com/orgs/openconfig/projects/2/views/1?filterQuery=)
+item. For example:
+
+* `tests/interfaces/rt_5_1_singleton_test.go` implements issue **RT-5.1
+  Singleton Interface**.
+* `tests/interfaces/rt_5_2_aggregate_test.go` implements issue **RT-5.2
+    Aggregate Interface**.
+
+## Pull Request Title
+
+A PR title should be as descriptive as possible,
+preferably starting with the scope (i.e. section in the test plan).
+
+```{.good}
+zz_1_1 testbed topology, no traffic generation yet
+```
+
+The PR description may add more details as desired to benefit the
+reviewers.  The preferred format is:
+
+```{.good}
+    * (M) yang/keychain/openconfig-keychain.yang
+      - Add a typedef for referencing a keychain from other modules.
+    * (M) yang/isis/*
+      - Fix support for hello-authentication to allow for references to a
+        specific keychain as defined in the keychain model.
+      - Fix support for authentication of *SNP packets, referencing a
+        keychain that can be used to auth these packets.
+      - move IS-IS model to openconfig-inet-types rather than ietf-inet-types.
+```
+
+## Code Organization
+
+Each code file should follow this general structure:
+
+```{.todo}
+TODO: Add preferred code structure
+```
+
+## IP Addresses Assignment
+
+Netblocks used in the test topology should follow IPv4 Address Blocks Reserved
+for Documentation ([RFC 5737]) and IPv6 Address Prefix Reserved for
+Documentation ([RFC 3849]). In particular:
+
+[RFC 5737]: https://datatracker.ietf.org/doc/html/rfc5737
+[RFC 3849]: https://datatracker.ietf.org/doc/html/rfc3849
+
+IPv4:
+
+* `TEST-NET-1`: 192.0.2.0/24
```

Comment:
I propose that we further restrict the assignment to be:

> * `TEST-NET-1` (192.0.2.0/24): control plane addresses split into /30 subnets for each ATE/DUT port pair.
> * `TEST-NET-2` (198.51.100.0/24): data plane source network addresses used for traffic testing; split as needed.
> * `TEST-NET-3` (203.0.113.0/24): data plane destination network addresses used for traffic testing; split as needed.

This is already de-facto in many tests. The rest of the tests will also follow this after they are converted to use the topology helper.